### PR TITLE
Bridge Dynamic to Static Type System 

### LINKED ISSUE #28: Type System
The operator code uses switch blocks in several places to properly handle different data types. We should have a type system that [replaces conditionals with polymorphism](https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html) (also see [this repo](https://github.com/UWQuickstep/gift)) or we could use templates.

### PR DESCRIPTION
**Pull request by GindaChen**
**(Draft)** Fixes #28. 

I chose the type trait + macro  approach to fill in the gap between static and dynamic typing system in arrow. 

### Usage

The usage is simple:
1. Use macro `HUSTLE_SWITCH_ARROW_TYPE` to translates an enum (dynamic) `DataType` type (enum value) to its corresponding `DataType` subclass. 
2. Use macro `HUSTLE_ARROW_TYPE_CASE_STMT` to define the case body. This need to be defined (and undefined) at each scope where `HUSTLE_SWITCH_ARROW_TYPE` is used - otherwise, by default, it throws an unimplemented error.
3. Define some templated function in the `STMT` macro, and use the `arrow_data_type` as a normal C++ type. This type is then used to instantiate template functions. 
4. For different types, user may want to handle them differently (e.g. whether or not it has native c-type, is default constructable, etc.) In this case, use **type traits** to put constraints on template init. 

See `type_helper.h` for details and example usage. Write-up and detailed design documentation (with its advantages, consequences, and best practices) comes later.

### Best practice for template function that use type traits (in C++17)

Type trait is a very hard topic in C++17. But we may only need to understand some of the common design patterns to navigate us in Hustle. This is one of the template design we commonly used:
```C++
// Common Usage
template<typename ArrowDataType, typename other-template-params...>
std::enabled_if<boolean_condition, function_return_type> foo(...) {}


// Simple Example: 
template<typename ArrowDataType>
std::enable_if_t<arrow::has_c_type<T>::value, arrow::Datum> foo (){...}
```

If you're confused with the example above, see this semantically equivalent example (but don't recommend using it):
```C++
// Declare the return type
using return_type = arrow::Datum

// Declare what the type T needs to satisfied. This "function" 
// (Actually a static variable inside a struct template instantiation, arbitrate by the SFINAE rule)
template<typename T>
using has_c_type_bool_result = arrow::has_c_type<T>::value;

// "Returns" the `return_type` if the first boolean is satisfied. 
// In this case, if has_c_type_bool_result<T> is true, then return `return_type`
// Otherwise, this template will not be instantialized.
template<typename T>
using return_type_if_have_ctype = std::enabled_if_t<has_c_type_bool_result<T>,  return_type>

// Contrast to the `return_type_if_have_ctype`. 
// This only instantiate when `!has_c_type_bool_result<T>` is satisfied.
template<typename T>
using return_type_if_not_have_ctype = std::enabled_if_t<!has_c_type_bool_result<T>,  return_type>

// Finally this function template
// This template only instantiate when has_c_type_bool_result<ArrowDataType> is true. 
// The function return type is `ArrowDataType`
template<typename ArrowDataType> 
return_type_if_have_ctype<ArrowDataType> foo(){...}

// If the function above is not instantialized, this will be instantiate. 
// This function make sure every arrow type have one and only one function to instantiate.
// Therefore, no overload conflict can happen. 
template<typename ArrowDataType> 
return_type_if_not_have_ctype<ArrowDataType> foo(){...}
```

**References**

- [SFINAE and enable_if](https://eli.thegreenplace.net/2014/sfinae-and-enable_if/). The two major concepts of type traits and why we can use the pattern above.
- [Template Metaprogramming: Type Traits (part 1 of 2) - Jody Hagins - CppCon 2020](https://www.youtube.com/watch?v=tiAVWcjIF6o)
- [Back to Basics: Templates (part 1 of 2) - Andreas Fertig - CppCon 2020](https://www.youtube.com/watch?v=VNJ4wiuxJM4)
- [CppCon 2014: Walter E. Brown "Modern Template Metaprogramming: A Compendium, Part I"](https://www.youtube.com/watch?v=Am2is2QCvxY)

### CODE REVIEW

**Comment by GindaChen** [comment]
@srsuryadev @mprammer2 This is a proposal to use  macro + template + type trait to solve the arrow type switch problem. Take a look and see if you have any questions, comments and suggestions

File: src/operators/expression.cc [hunk 1]
```diff
@@ -121,6 +122,37 @@ arrow::Datum Expression::ExecuteBlock(bool is_result,
   }
 }
 
+template <typename T>
+std::enable_if_t<arrow::has_c_type<T>::value, arrow::Datum>
+Expression::ExecuteBlockHandler(
+    bool is_result, int op, const std::shared_ptr<arrow::Array>& left_col,
```
[the remark this answers is not in the document]
Comment (GindaChen) [reply src/operators/expression.cc:128]: I will talk about this in our tomorrow meeting. There is a simpler way to do this, and I just did it here:
https://github.com/UWHustle/hustle/pull/94/files#diff-9e0d518effb81ef7dd04c1a1966975f9907f06f10632bbda2c4038dfb5e5e8d2R374-R415

Here we write it in **waterfall**: For each groups of type that satisfies certain traits, we direct them to a function asap. I am refactoring this piece in a similar way. Wait and see!

**Comment by GindaChen** [comment]
Build pass on local. But the `template` specialization seems to have some problem to compile in the CircleCI.

**Comment by johnchronis** [comment]
very nice PR and description @GindaChen! Once this is done you should teach how to use the new C++17/20 template features :)

**Comment by GindaChen** [comment]
If people still wants to come to Friday meeting, I can plan for an intro to that feature : )

**Review by mprammer2 — APPROVED** [review]
Approved based on today's meeting with @GindaChen and @srsuryadev .

**Comment by GindaChen** [comment]
Thanks @mprammer2 and @srsuryadev! Will make some minor changes on docs and aliases before closing.

**Comment by GindaChen** [comment]
@mprammer2 @srsuryadev FYI - Just took another look at the arrow code base and found some inspiration through their `Visitor`.

https://github.com/apache/arrow/blob/2d140c3eeecca3ff7823edc8c9562ebd6a1c336a/cpp/src/arrow/visitor_inline.h#L37-L51

This way, they seem to pass in the `ACTION` (i.e. our case statement) code block directly into the case statement. I'm trying to see if it works - if it does work, it could simplify our redefine-macro approach. 

To answer @srsuryadev's question about why (one way of) polymorphism can't work / is too tedious to work, we can look at the visitor (virtual base) class implementation: 
https://github.com/apache/arrow/blob/master/cpp/src/arrow/visitor.h

This will work for every type iff we override every function within this visitor. For example, to `getBuilder`, we don't have the option to **group** the types and handle them all together -- because it is not compile-time selection of type. What we can only do, at best, is to define the same template function like we did, and then (follow the virtual function overload guide) put the template function -- and instantiate the type -- within each override function body. Oh no! 30 types handwritten! Even if we use the `ACTION` macro trick in the [visitor-inline.h](https://github.com/apache/arrow/blob/2d140c3eeecca3ff7823edc8c9562ebd6a1c336a/cpp/src/arrow/visitor_inline.h#L37-L51), it is of no help because we really want to **do something for each group of types**. 

---

To conclude:
1. The (rewrite) macro + template traits is by far the best and flexible solution to solve the typing problem. 
2. This is a steeper hill that people have to climb, and I'd admit it's our job to mask some of the complexity by providing standard design patterns in hustle for developer and help them program with less overhead as possible. 
3. Meanwhile, I'm experimenting at how C++20 will help us solve the problem more elegantly (hopefully not adding overhead). If C++20 can solve our problem, we may want to bump up and enjoy life in the whole new world..

**Pull request merged**